---
title: "Human Genome Analysis Lab 4 : Publication Graphics"
output:
  html_document:
    toc: true
    toc_depth: 4
    theme: united
    highlight: tango
editor_options: 
  chunk_output_type: console
---

## Learning objectives

* Be able to add title, axis labels, legends, colors to ggplot graphs
* Resize graphs in RMarkdown
* Print graphics to a file (e.g. jpeg, pdf)
* Loading images into a RMarkdown file
* Making interactive graphs and tables in RMarkdown

## Fine tuning ggplots

Today we will use the <a href="http://www.cookbook-r.com/Graphs/">Cookbook for R by Winston Chang</a> for tidying up our graphs. Please take your time and go through the following web pages. Copy and paste the example code into R Studio. 

* <a href="http://www.cookbook-r.com/Graphs/Titles_(ggplot2)/">Cookbook for R - Title</a>
* <a href="http://www.cookbook-r.com/Graphs/Axes_(ggplot2)/">Cookbook for R - Axes</a>
* <a href="http://www.cookbook-r.com/Graphs/Legends_(ggplot2)/">Cookbook for R - Legends</a>
* <a href="http://www.cookbook-r.com/Graphs/Colors_(ggplot2)/">Cookbook for R - Colors</a>

Here are a couple of cheatsheets that can be useful

* <a href="http://www.rstudio.com/wp-content/uploads/2015/12/ggplot2-cheatsheet-2.0.pdf">R Studio ggplot2 cheatsheet</a>
* <a href="https://rstudio.com/wp-content/uploads/2016/03/rmarkdown-cheatsheet-2.0.pdf">RMarkdown cheatsheet</a>

### Controlling graph size in RMarkdown 

In the opening line of the RMarkdown code chunk {r} you can control the output of the code, graphs, tables using knitr syntax. For example if {r, eval = FALSE} the code will not be run, but will be shown. If {r, code = FALSE} the code will not be shown, but will be run and the output will be shown (useful in reports where the reader is only interested in the results/graphs, but not the code). You can also suppress error messages and warnings so that the reader isn't bothered by them (but you should take notice).

The dimensions of an individual graph in the RMarkdown document be adjusted by specifying the graph dimensions as below.

```{r, message = FALSE}
# Load Libraries
library(tidyverse)
```

```{r}
# Load Files 
SNPs<- read.table("data/23andMe_complete.txt", header = TRUE, sep = "\t")
```

```{r, fig.width = 6, fig.height = 6}
# to adjust figure size {r, fig.width = 6, fig.height = 6}
SNPs$chromosome = ordered(SNPs$chromosome, levels=c(seq(1, 22), "X", "Y", "MT"))
ggplot(data = SNPs) + 
  geom_bar(mapping = aes(x = genotype, fill = chromosome)) + 
  coord_polar() +
  ggtitle("Total SNPs for each genotype") +
  ylab("Total number of SNPs") +
  xlab("Genotype")
```

### Graphic Output

You may have realized that you can export plots in R Studio by clicking on Export in the Plots window that appears after you make a graph. You can save as a pdf, svg, tiff, png, bmp, jpeg and eps.  You can also write the output directly to a file. This is particularly useful for controling the final dimensions in a reproducible way and for manuscripts.

* <a href="http://www.cookbook-r.com/Graphs/Output_to_a_file/">Cookbook for R - Output to a file - PDF, PNG, TIFF, SVG </a>

```{r}
# Plot graph to a pdf outputfile
pdf("images/SNP_example_plot.pdf", width=6, height=3)
ggplot(data = SNPs) + 
  geom_bar(mapping = aes(x = chromosome, fill = genotype))
dev.off()
```

```{r}
# Plot graph to a png outputfile
ppi <- 300
png("images/SNP_example_plot.png", width=6*ppi, height=6*ppi, res=ppi)
ggplot(data = SNPs) + 
  geom_bar(mapping = aes(x = chromosome, fill = genotype))
dev.off()
```

For more details on sizing output <a href="http://www.cookbook-r.com/Graphs/Output_to_a_file/">Cookbook for R - Output to a file - PDF, PNG, TIFF, SVG </a>

### RMarkdown loading images

Sometimes it is useful in controling the image layout for a report to file with the graph and then subsequently load it into the .Rmd file. This works with png files, but not pdfs. You can also upload images made with other bioinformatic tools into your RMarkdown report.

```{r eval = FALSE}
# This is the RMarkdown style for inserting images
# Your image must be in your working directory
# This command is put OUTSIDE the r code chunk

![Genotype counts per chromosome](images/SNP_example_plot.png) 
```

![Genotype counts per chromosome](images/SNP_example_plot.png) 

```{r eval = FALSE}
# This is an alternative way using html. 
# Remember that it must be in your working directory or you will need to specify the full path.
# The html is put OUTSIDE the r code chunk.

<img src="images/SNP_example_plot.png" alt="Genotype counts per chromosome" style="width: 600px;"/>
```

<img src="images/SNP_example_plot.png" alt="Genotype counts per chromosome" style="width: 600px;"/>

Another way to present a graph without the code is adding echo = FALSE within the r{} chunk - {r echo = FALSE}. This prevents code, but not the results from appearing in the knitr file.

### Interactive graphs and tables in RMarkdown reports

With plotly/ggplotly (https://plot.ly/ggplot2/) you can make interactive graphs in your lab report.

```{r, message=FALSE, eval = FALSE}
# Version 1 1
library(plotly)
p <- ggplot(data = iris, aes(x = Sepal.Length, y = Sepal.Width, color = Species)) + 
  geom_point()
ggplotly(p)
```


```{r, message=FALSE}
# Version 2
library(plotly)
ggplotly(
  ggplot(data = iris, aes(x = Sepal.Length, y = Sepal.Width, color = Species)) + 
    geom_point()
 )
```

You can also make interactive data tables with the DT package (https://rstudio.github.io/DT/)
*** Don't do this with tables of hundreds of thousands of rows (as in your complete SNP table)

```{r}
library(DT)
datatable(iris)
```

## Exercises


### Exercise 1

Add title and labels for the x and y axis to Lab3 ex1. Color the bars blue

```{r}
ggplot(SNPs) + aes(chromosome) + geom_bar(fill = "blue") + labs(title="Number of positions per chromosome", y="Number of positions", x="Chromosome")

```


### Exercise 2

To Lab3 ex3 add more defined x and y axis labels, add a title, Change the colors of the genotypes, so that the dinucleotides (e.g. AA) are one color, the mononucleotides (A) are another and D’s and I’s a third color. One way to do this is to specify the color of each genotype.

```{r}
cols <- c("A" = "yellow", "AA" = "red", "AG" = "red", "AC" = "red", "AT" = "red", "CG" = "red", "CT" = "red", "GT" = "red", "DI" = "blue", "D" = "blue", "DD" = "blue", "I" = "blue", "II" = "blue", "G" = "yellow", "T" = "yellow", "C" = "yellow", "GG" = "red", "CC" = "red", "TT" = "red", "--" = "white")
ggplot(SNPs) + aes(chromosome, fill = genotype) + geom_bar() + labs(title="Contribution of genotypes to positions per chromosome", x="Chromosome", y="Number of positions") + scale_fill_manual(values = cols)
```

### Exercise 3

From Lab3 ex5, make an output png file, then load the file into report using the RMarkdown or html format. 

```{r}
ppi <- 300
png("images/SNP_example_plot2.png", width=6*ppi, height=6*ppi, res=ppi)
ggplot(SNPs) + aes(chromosome, fill = genotype)+
  geom_bar(position = "dodge")
dev.off()

```
![Genotype counts per chromosome](images/SNP_example_plot2.png) 

### Exercise 4

For Lab3 ex6 add more descriptive x and y axis labels, add a title, make the x-axis for each graph readable in your final report file. 
```{r}
ggplot(SNPs) + aes(chromosome, fill = genotype) +
  geom_bar(position = "dodge") + facet_wrap(~genotype) + labs(title = "Genotype contributions to chromosome count") + theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))
```

### Exercise 5

Turn Lab3 ex6 into an interactive graph using plotly

```{r}
ggplotly(
  ggplot(SNPs) + aes(chromosome, fill = genotype) +
  geom_bar(position = "dodge") + facet_wrap(~genotype) + labs(title = "Genotype contributions to chromosome count") + theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5))
)
```


### Exercise 6

Make an interactive table of the SNPS found in on the Y chromosome from the 23andMe_complete data set
```{r}
library(dplyr)
datatable(filter(SNPs, chromosome == "Y"))
```

